### ISSUE TITLE
RFC: StoreV2 deprecation plan

### ISSUE THREAD

**Issue by ptabor** [etcd-io/etcd#12913]
[The content of this post is being edited, and is not yet approved plan of record ]

# Background

In 3.4 storeV2 is still extensively used: 
  - User can opt in `--enable-v2` to unable V2 API that writes solely data to storeV2 files. 
  - Membership information is both stored in V2 & V3 (backend) stores  
  - Membership information is being read (recovered) from storeV2
  - Publishing membership happens through StoreV2 raft operation.
     - The new ClusterMemberAttrSet applier was implemented in etcd 3.5 so cannot be used by default in 3.5: to allow 3.4->3.5 rollback.   
  - During startup etcd assumes each WAL log snapshot is accompanied by storev2 snapshot. 

- Note: `etcdctl snapshot restore ` is not restoring V2 content   (producing fake storeV2 with membership information)

# Plan: 

## 3.5 release (updated by ahrtr )

Both V2 and V3 state are supported, but V3 becomes the source of truth. TODO:
 - [x] --experimental-enable-v2v3= promoted to `--enable-v2v3=...` or deprecated. [see https://github.com/etcd-io/etcd/issues/12905]
 - [x] --enable-v2v3=...  should allow mounting as root ?
 - [x] --enable-v2 prints deprecation warning [optional: when not combined with `--enable-v2v3=...` ]
 - [x] ClusterMemberAttrSet is implemented. 
 - [x] Membership is only read from backend v2  (https://github.com/etcd-io/etcd/pull/12914)
 - [x] --experimental-no-v2=PHASE_1_WRITE_ONLY mode that: 
     - cannot be used when `--enable-v2` is ON
     - if the storev2 snapshot is found accompanying the (used to be used) WAL log snapshot, 
       its validated whether it has no used-data (apart of membership & version)
     - storev2 is content created on demand using 'etcdctl snapshot restore' logic based on the last WAL snapshot 
 - [x] https://github.com/etcd-io/etcd/issues/18993  
 - [x] https://github.com/etcd-io/etcd/pull/18999
## 3.6 release (updated by serathius@, ahrtr )

V3 is the only meaningful state. V2 state is generated from V3 state to maintained backward compability.  TODO:
- [x] `--enable-v2` is docomissioned
- [x] PublishV3 is used solely
- [x] Remove v2 commands from etcdctl (https://github.com/etcd-io/etcd/pull/13643)
- [x] Remove v2 client library (https://github.com/etcd-io/etcd/pull/15591)
- [x] Discovery v3 is implemented https://github.com/etcd-io/etcd/issues/13624
- [x] experimental-no PHASE_1_WRITE_ONLY becomes the default.
- [x] Discovery v2 is deprecated
- [x] Tests for backward v2 compatibility are implemented https://github.com/etcd-io/etcd/pull/13756
- [x] V2 snapshots are generated from V3 data
- [x] V2 apply code is removed
- [x] ~~V2 no longer stores membership data~~
- [x] Cleanup V2 code to leave only code for generating snapshots.
- [x] Decide on https://github.com/etcd-io/etcd/issues/17009 and implement --v2-deprecation=write-only-drop-data
- [x] https://github.com/etcd-io/etcd/issues/18994
- [x] https://github.com/etcd-io/etcd/issues/18995
- The document will be covered in https://github.com/etcd-io/website/issues/926
 
 ## 3.7 release (updated by serathius@, ahrtr )
 
 - [ ] V2 code is totally removed. TODO:
   - [x] Discovery v2 is removed
   - [x] client v2 is removed
   - [ ] All V2 store code is removed
   - [ ] Only load data (i.e. members) from v3store (bbolt)
   - [ ] Validate membership applying using v3store
 - [ ] Do not generate v2 snapshot files at all
 - [ ] Bootstrap etcdserver from v3store(bbolt) instead of v2 snapshot: https://github.com/etcd-io/etcd/issues/20187 
 - [ ] Change the default value for `--v2-deprecation` to [V2Depr1WriteOnlyDrop](https://github.com/etcd-io/etcd/blob/2abb334103a327270e4812fa0da21e941125440f/server/config/v2_deprecation.go#L39)
 - [ ] rewrite `TestV2DeprecationWriteOnlyWAL` and `TestV2DeprecationWriteOnlySnapshot`, refer to https://github.com/etcd-io/etcd/pull/19961

## 3.8 release
Remove `--v2-deprecation`. Completely cleanup anything related to v2store.

## Docs
- https://docs.google.com/document/d/14PYLcdTyX3jhdBDc6q6WWRceTaIZePauHTWgjHJ0_ZE/edit?tab=t.0
- ~~https://docs.google.com/document/d/1dpvBrOvBquZ8QozJcT6xcwVUVXYwO3hvjmVBd7ixIJU/edit~~

**Comment by xiang90** [comment]
@ptabor 

Thanks for helping with this! (I wanted to have a clean separation between v2 and v3 a couple of years ago but never got time actually to do it).

I am not sure if we are still running the failure injection tests. I would suggest running them to ensure the correctness of this change. Also, we might want more tests on the membership changes and the upgrade path.

/cc @gyuho

**Comment by ptabor** [comment]
@xiang90 
The functional tests (https://travis-ci.com/github/etcd-io/etcd/jobs/502908317) looks to me to be a failure injection tests in practice. 
Do you mean these tests, or there are other tests ?

**Comment by xiang90** [comment]
@ptabor 

The functional test randomly injects failures into the running servers. Due to the randomness, it might not be super effective to catch issues if it only runs one round in the CI. 

What we did previously was to run this test 24*7 for weeks on a set of servers before major releases. I am not sure if it is still the practice since we have not really introduced significant changes to etcd in the recent releases I believe.

**Comment by gyuho** [comment]
@ptabor This looks great.

Regarding

>  --experimental-enable-v2v3= promoted to --enable-v2v3=... or deprecated

as we discussed, let's mark it as deprecated for 3.5 and remove for 3.6.

Then, there's no need for

> --enable-v2 is docomissioned or FAILs if not accompanied by --enable-v2v3=...

?

**Comment by ptabor** [comment]
I hope/think mandatory steps for v3.5 are done.

**Comment by serathius** [comment]
I can take a look deprecating V2

**Comment by der-eismann** [comment]
Not sure if this is the right place to ask, but when starting the logs say
```
detected custom v2store content. Etcd v3.5 is the last version allowing to access it using API v2. 
**Please remove the content.**
```

But I couldn't find how to do that. I removed all keys from the v2 store, but apparently there is still something stored in it. How do I remove all v2 content?

**Comment by serathius** [comment]
Updated the top comment with the plan based on my discussion with @ptabor and work on deprecation. One difference is that I'm proposing to provide v3.6->v3.5 backward compatibility by using V3 state to generate V2 snapshots. This of cause needs proper testing,  I have implemented some tests for this https://github.com/etcd-io/etcd/blob/main/tests/e2e/v2store_deprecation_test.go, but would be good to double check.

If someone is interested in helping, please let me know.

**Comment by geetasg** [comment]
I am interested in working on this issue. Investigating changes.

**Comment by ahrtr** [comment]
> provide v3.6->v3.5 backward compatibility by using V3 state to generate V2 snapshots

Please clarify this to avoid any confusion or misunderstanding. Are the "`V2 snapshots`" the `*.snap` files under the `${ETCD_DATA_DIR}/member/snap` directory?

**Comment by serathius** [comment]
> > provide v3.6->v3.5 backward compatibility by using V3 state to generate V2 snapshots
> 
> Please clarify this to avoid any confusion or misunderstanding. Are the "`V2 snapshots`" the `*.snap` files under the `${ETCD_DATA_DIR}/member/snap` directory?

Yes, V2 snapshots are the `*.snap` files.

**Comment by ahrtr** [comment]
> provide v3.6->v3.5 backward compatibility by using V3 state to generate V2 snapshots

It is still not clear what does it exactly mean.

**Comment by serathius** [comment]
Replace code that generates V2 snapshots, and modify it so it doesn't use v2 store, but v3 store.

**Comment by ahrtr** [comment]
> Replace code that generates V2 snapshots, and modify it so it doesn't use v2 store, but v3 store.

It's a big change. Have we considered completely removing the v2 snapshot files as well in 3.7?  If yes, then the other option is to keep it as it's in 3.6, and remove both v2store and v2 snapshot files in 3.7. We can get all information form bbolt db, why do we need separate v2 snapshot files?

I am not saying which approach we must follow. Instead, I mean it's a big change, we can't simply to make change just based on one comment. Please evaluate the impact of all possible approaches and related effort.

**Comment by serathius** [comment]
> It's a big change. Have we considered completely removing the v2 snapshot files as well in 3.7?

It's a big change in sense how much it would simplify apply logic and number of lines of code could remove. However, it's not big change in sense of being hard to implement. Compared to most of the code/features, backward compatibility is much easier to test. Please check https://github.com/etcd-io/etcd/pull/13756

> We can get all information form bbolt db, why do we need separate v2 snapshot files?

Backward compatibility with v3.5. 

>  we can't simply to make change just based on one comment. Please evaluate the impact of all possible approaches and related effort.

It was not based on one comment. I have been working on removal of v2store for over a year. Sorry that not everything was documented back then, but I was the only one doing the work. Removing v2store early should have huge impact 
on simplifying the codebase and increasing project velocity. I think it's worth to do more for v3.6 release vs wait another 2 years for v3.7.

**Comment by ahrtr** [comment]
> It was not based on one comment. I have been working on removal of v2store for over a year. 

I completely agree with the direction of removing v2store. I am just concerned about how was the decision "`using V3 state to generate V2 snapshots`" made. I do not see any detailed discussion or doc on that. Also we might also need to consider to completely remove v2 snapshots in future, but I also do not see any such discussion or items in the Plan.

**Comment by serathius** [comment]
Adding my old attempt to implement this https://github.com/etcd-io/etcd/commit/2ddb174bc35d829a798310a27fc4ec8923a506bf

**Comment by geetasg** [comment]
Link to my current attempt - https://github.com/etcd-io/etcd/pull/15905
This borrows from https://github.com/etcd-io/etcd/commit/2ddb174bc35d829a798310a27fc4ec8923a506bf and also has code to export v2 snapshots for backward compatibility

**Comment by geetasg** [comment]
Related doc  - https://docs.google.com/document/d/19_7xBwZBNrM-75h5rr2PUWFM_eNjSXRVPZZpVaTS2qw/edit?usp=sharing

**Comment by geetasg** [comment]
The PRs will be roughly in following order -

- [ ] Update/add tests
- [ ] Introduce method and its consumers to publish v2 snapshot from v3 state.Ref:https://github.com/etcd-io/etcd/pull/16000/files#diff-09b9c924aaa34d5ad5abc7e40ed10a2dcc5257c62fda537a8671a7f206d03e6aR831
- [ ] Stop / remove use of SetStore method of RaftCluster in bootstrap. There should not be any caller of RaftCluster SetStore at this point.
- [ ] Remove v2 store from RaftCluster
- [ ] Introduce applier_v2v3 and tests
- [ ] Remove v2store from EtcdServer
- [ ] Remove v2 applier
- [ ] Update --experimental-no-v2=PHASE_1_WRITE_ONLY  to write_only_drop_data
- [ ] Update docs about write_only_drop_data

**Comment by geetasg** [comment]
@ahrtr Thank you for the review.
Plan for "V2 snapshots are generated from V3 data"  is listed below -
1.Provide a method to export membership in v2 format - Ref:https://github.com/etcd-io/etcd/pull/16132
2.Consume the method implemented in above step from triggerSnapshot workflow as in poc here - https://github.com/geetasg/etcd/blob/poc1/server/etcdserver/server.go#L2055
3.Similarly consume the same method from etcdutl backup as shown here - https://github.com/geetasg/etcd/blob/poc1/etcdutl/etcdutl/backup_command.go#L184
4.Consume the method from etcdutl snapshot
Ref:https://github.com/geetasg/etcd/blob/poc1/etcdutl/snapshot/v3_snapshot.go#L461
5.Consume the method from workflow to created merged snapshot Ref:https://github.com/geetasg/etcd/blob/poc1/server/etcdserver/snapshot_merge.go#L34

This poc is at branch poc1 in my repo. 
Please let me know if you have questions/suggestions on this plan. I am currently at step 1 - PR is published for review.  
/cc @serathius @ptabor

**Comment by geetasg** [comment]
Also listing the plan for other items -
V2 no longer stores membership data
1.Fix conf change entry validation - https://github.com/geetasg/etcd/tree/pr3
2.Remove v2store from RaftCluster - ref:https://github.com/etcd-io/etcd/pull/16000/commits/8dcc34d5f2609e6d223ebfd58225a91d1a54a9be - some of the earlier commits leading upto this removal will count too.
3.Remove v2store from bootstrap - Ref: https://github.com/etcd-io/etcd/pull/16000/commits/930ecd28bbd4dc63b89aff26e28e547ec5b056a6
and https://github.com/etcd-io/etcd/pull/16000/commits/9b0dd9760a81a5b9d8a6ed009f3d668dc7350ef5

Currently this work is on #1 - PR is being discussed..

**Comment by geetasg** [comment]
Plan for V2 apply code is removed -
1.This will involve introducing a new applier which can apply v2 req to v3 backend. Ref:https://github.com/etcd-io/etcd/pull/16000/commits/733d109241bcf60c590fce9fd04f1e98d710dbfa
2.Update tests - there are unit tests in server_test.go that issue v2 requests. Ref: TestDoLocalAction
3.Remove v2 applier - https://github.com/etcd-io/etcd/pull/16000/commits/9c02d763d4b9859b7a648f3fd1050d126e84fcc1

v2v3 applier can be removed once the 3.5 publish node request is n/a.

**Comment by geetasg** [comment]
>>Remove the etcdutl backup? 
makes sense to me. It can be removed if there are no objections.

**Comment by geetasg** [comment]
@ahrtr @serathius @ptabor please let me know if you have any questions/comments on the plan listed above. These changes are poced in https://github.com/etcd-io/etcd/pull/16000. Thanks

**Comment by ahrtr** [comment]
@geetasg  Could you summarize all remaining tasks in one comment using a hierarchical structure to clearly show what has been completed and what has not? thx

**Comment by geetasg** [comment]
@ahrtr updated https://github.com/etcd-io/etcd/issues/12913#issuecomment-1593431504 above to indicate latest status. 

In terms of high level tasks as listed in Plan section above for 3.6 release - 

- [ ] V2 snapshots are generated from V3 data -- done
- [ ]  V2 apply code is removed
- [ ]  V2 no longer stores membership data -- WIP
- [ ]  Cleanup V2 code to leave only code for generating snapshots.

**Comment by ahrtr** [comment]
Done.

The decade-plus effort is done. v2store is finally completely gone!!!

See https://github.com/etcd-io/etcd/issues/20187#issuecomment-5442917035
